tests/_helpers/setup.ts

```diff
@@ -28,24 +28,26 @@ export function render<Elements extends Element | Element[] = HTMLElement>(
   if (typeof focus === 'string') {
     ;(assertSingleNodeFromXPath(focus, div) as HTMLElement).focus()
   } else if (focus !== false) {
-    ;(div.querySelector(FOCUSABLE_SELECTOR) as HTMLElement | undefined)?.focus()
+    const element: HTMLElement | null = div.querySelector(FOCUSABLE_SELECTOR)
+    if (element) {
+      element.focus()
+    } else if (div.firstChild) {
+      ;(div.firstChild as HTMLElement).focus()
+    }
```

Comment:
No, I assumed it is a custom element. I changed this to check for firstElementChild and check the tag for `hyphen`, thus making clear it is a custom element and therefore focusable.

Reply:
Not every custom element is focusable. And the behavior depends on `ShadowRoot.delegatesFocus` or the overloaded `HTMLElement.focus`.
This util is supposed to provide a good default for our tests that makes it easy for future developers to write tests that focus on the tested feature without polluting the tests with too much arrange logic.
I want to make sure there is nothing subtle about this function. It currently focuses the first focusable element in the rendered fragment by default.
Let's add a function that ensures this behavior when one or more elements with shadow DOM are used here.

Reply:
At work we only have custom elements that are focusable. I just tried this in a variety of browsers and got different results per browser. I am super confused now ;) I will try to write a little helper, so we can figure out the focusable element and call it a day.

Reply:
```ts
function findFocusable(container: Element | ShadowRoot) {
  for(const el of container.querySelectorAll('*')) {
    if (el.matches(FOCUSABLE_SELECTOR)) {
      return el
    } else if (el.shadowRoot) {
      const f = findFocusable(el.shadowRoot)
      if (f) {
       return f
      }
    }
}
```

Reply:
I added a helper myself which also skips searching inside the shadowRoot if delegatesFocus is defined.

Reply:
`ShadowRoot.delegatesFocus` can be `false`.
`ShadowRoot.delegatesFocus` can be `true`, but the element could not contain a focusable element.
A focusable element inside a shadow tree might precede or follow on a focusable element outside of it.

Reply:
This is true. My idea was to fallback to the engine's implementation if there is one. I think overriding this behavior in general if it exists in the browser engine is a bad idea. It is true though, we will not know what the element will do with the focus.

Reply:
Done

Reply:
The goal is to provide a state to the document that can be used to test a certain feature.
The util needs to provide the same state in different environments.
We will probably run these tests in Jsdom and in a browser. See #1019

Reply:
I am not convinced. This feels like reimplementing the browser. If I want to test pasting support in my custom element for example, I rely on the proper element being focused. What the browser does on `delegatesFocus` is important to get the proper element focused. 

I may want to use karma to run this test in a real browser and I would assume that `delegatesFocus` would be used, because that's why I am running it in a real browser. While in reality what would happen would be that `getFocusableElement` would be used, yet as a user I would be unaware of this. If our implementation differs from the browser's implementation people would never notice this, because we would always use our implementation without them knowing.

Reply:
> If I want to test pasting support in my custom element for example, I rely on the proper element being focused. What the browser does on delegatesFocus is important to get the proper element focused.

That's exactly the issue here. The setup util provides a certain document state, e.g. `elementA` being focused, so that a test can assert a feature is correctly implemented, e.g. the pasted content being inserted in `elementA`.
If the setup util provides a different document state, e.g. `elementB` being focused, it is expected that pasted content is not being inserted in `elementA`.

Now that `elementA` has to get focused on some interaction is a different feature. On that one it is important that there is a certain document state, e.g. `elementA` not being focused.

If the document state created by the setup util depends on the environment, this might make it very difficult to find the relevant bit when multiple tests break and point at different features.

Reply:
Oh, sorry, I was mistaken. `setup` is only really used internally right? Then I agree this is a good idea. Sorry for the confusion. I updated the code.

Reply:
Yes, it's part of the `#testHelpers`. When we implement handling focus on the `mousedown` on the shadow tree, we'll need to provide the same behavior as the browser would.